File: scripts/sync-translations.ts
#!/usr/bin/env node
/**
 * Sync Translations - Append missing strings to the EN JSON files.
 *
 * Runs the full code scan and appends any strings found in code but
 * absent from the EN JSON files. Menu items (controls prefixed `mnu*`)
 * go to r_instat_menus.json; everything else goes to r_instat_not_menus.json.
 *
 * Usage:
 *   npm run sync-translations              # Append missing, write files
 *   npm run sync-translations:dry-run      # Print what would be added, no writes
 *
 * Options:
 *   --dry-run     Report changes without writing files
 *   --verbose     Show detailed scan logs
 *   --help        Show help
 */

import { getProjectRoot } from './lib/project-root.js';
import { runScan } from './lib/scan.js';
import { appendMissingStrings } from './lib/translation-writer.js';

interface CliOptions {
  dryRun: boolean;
  verbose: boolean;
  help: boolean;
}

function parseArgs(args: string[]): CliOptions {
  const options: CliOptions = { dryRun: false, verbose: false, help: false };
  for (const arg of args) {
    switch (arg) {
      case '--dry-run':
        options.dryRun = true;
        break;
      case '--verbose':
      case '-v':
        options.verbose = true;
        break;
      case '--help':
      case '-h':
        options.help = true;
        break;
    }
  }
  return options;
}

function printHelp(): void {
  console.log(`
Sync Translations - Append missing strings to the EN JSON files.

Usage:
  npx ts-node sync-translations.ts [options]

Options:
  --dry-run     Report changes without writing files
  --verbose     Show detailed scan logs
  --help        Show this help message

Menu items (controls prefixed mnu*) go to r_instat_menus.json.
All other strings go to r_instat_not_menus.json.
`);
}

function summarize(title: string, items: string[]): void {
  console.log(`${title}: ${items.length}`);
  const preview = items.slice(0, 5);
  for (const item of preview) {
    const truncated = item.length > 70 ? item.slice(0, 70) + '...' : item;
    console.log(`  + "${truncated}"`);
  }
  if (items.length > preview.length) {
    console.log(`  ... (${items.length - preview.length} more)`);
  }
}

async function main(): Promise<void> {
  const options = parseArgs(process.argv.slice(2));

  if (options.help) {
    printHelp();
    process.exit(0);
  }

  let baseDir: string;
  try {
    baseDir = await getProjectRoot();
  } catch (error) {
    console.error((error as Error).message);
    process.exit(1);
  }

  if (options.verbose) {
    console.log(`Base directory: ${baseDir}`);
    console.log(`Mode: ${options.dryRun ? 'dry-run' : 'write'}`);
  }

  const { classifiedMissing, stats } = await runScan(baseDir, {
    ciMode: false,
    verbose: options.verbose
  });

  console.log('');
  console.log('Scan results:');
  console.log(`  Files scanned:       ${stats.totalFilesChecked}`);
  console.log(`  Strings extracted:   ${stats.totalStringsExtracted}`);
  console.log(`  Strings ignored:     ${stats.totalStringsIgnored}`);
  console.log(`  Missing in EN JSON:  ${classifiedMissing.size}`);
  console.log('');

  if (classifiedMissing.size === 0) {
    console.log('✅ Nothing to sync — all extracted strings exist in EN JSON.');
    process.exit(0);
  }

  // Partition for reporting.
  const wouldAddMenus: string[] = [];
  const wouldAddNotMenus: string[] = [];
  for (const [key, route] of classifiedMissing.entries()) {
    if (route === 'menus') wouldAddMenus.push(key);
    else wouldAddNotMenus.push(key);
  }

  if (options.dryRun) {
    console.log('Dry-run — no files written.');
    console.log('');
    summarize('Would append to r_instat_menus.json', wouldAddMenus);
    console.log('');
    summarize('Would append to r_instat_not_menus.json', wouldAddNotMenus);
    process.exit(0);
  }

  const { added } = await appendMissingStrings(baseDir, classifiedMissing);

  console.log('✍️  Wrote EN JSON updates.');
  console.log('');
  summarize('Appended to r_instat_menus.json', added.menus);
  console.log('');
  summarize('Appended to r_instat_not_menus.json', added.notMenus);
  console.log('');
  console.log('Next: review the diff and commit the EN JSON changes along with your VB edits.');

  process.exit(0);
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});

File: scripts/sync_translations.py
#!/usr/bin/env python3
"""Sync translations entry point.

Appends missing strings from VB.NET scan results into the EN translation JSON
files under instat/translations/en/.
"""

from __future__ import annotations

import argparse
import sys

from remove_translation_strings import run_remove_translation_strings

from translation_common import (
    append_missing_strings,
    find_project_root,
    load_translation_file,
    MENUS_REL,
    NOT_MENUS_REL,
    remove_text_ignored_strings,
    run_scan,
    should_ignore_text,
    summarize_items,
)


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Sync Translations - Append missing strings to the EN JSON files"
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Report changes without writing files",
    )
    parser.add_argument(
        "--verbose",
        "-v",
        action="store_true",
        help="Show detailed scan logs",
    )
    parser.add_argument(
        "--remove_orphan_strings",
        action="store_true",
        help="Run remove_translation_strings as part of sync",
    )
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    options = parse_args(argv)

    try:
        base_dir = find_project_root()
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    if options.verbose:
        mode = "dry-run" if options.dry_run else "write"
        print(f"Base directory: {base_dir}")
        print(f"Mode: {mode}")

    scan_result = run_scan(base_dir, ci_mode=False, verbose=options.verbose)
    classified_missing = scan_result["classifiedMissing"]
    ignore_patterns = scan_result["ignorePatterns"]
    stats = scan_result["stats"]

    print("")
    print("Scan results:")
    print(f"  Files scanned:       {stats['totalFilesChecked']}")
    print(f"  Strings extracted:   {stats['totalStringsExtracted']}")
    print(f"  Strings ignored:     {stats['totalStringsIgnored']}")
    print(f"  Missing in EN JSON:  {len(classified_missing)}")
    print("")

    auto_removed = {"menus": [], "notMenus": []}
    if options.remove_orphan_strings:
        remove_result = run_remove_translation_strings(
            base_dir,
            dry_run=options.dry_run,
            verbose=options.verbose,
            scan_result=scan_result,
        )
        auto_removed = remove_result["removed"]
        print(
            summarize_items(
                "Auto remove_translation_strings from r_instat_menus.json",
                auto_removed["menus"],
            )
        )
        print("")
        print(
            summarize_items(
                "Auto remove_translation_strings from r_instat_not_menus.json",
                auto_removed["notMenus"],
            )
        )
        print("")

    if options.dry_run:
        menus_data, _ = load_translation_file(base_dir / MENUS_REL)
        not_menus_data, _ = load_translation_file(base_dir / NOT_MENUS_REL)
        would_remove_menus = [k for k in menus_data if should_ignore_text(k, ignore_patterns)]
        would_remove_not_menus = [k for k in not_menus_data if should_ignore_text(k, ignore_patterns)]

        would_add_menus = [k for k, v in classified_missing.items() if v == "menus"]
        would_add_not_menus = [k for k, v in classified_missing.items() if v != "menus"]

        print("Dry-run - no files written.")
        print("")
        if would_remove_menus or would_remove_not_menus:
            print(summarize_items("Would remove from r_instat_menus.json (text: rules)", would_remove_menus))
            print("")
            print(summarize_items("Would remove from r_instat_not_menus.json (text: rules)", would_remove_not_menus))
            print("")
        print(summarize_items("Would append to r_instat_menus.json", would_add_menus))
        print("")
        print(summarize_items("Would append to r_instat_not_menus.json", would_add_not_menus))
        return 0

    removed = remove_text_ignored_strings(base_dir, ignore_patterns)

    if removed["menus"] or removed["notMenus"]:
        print(summarize_items("Removed from r_instat_menus.json (text: rules)", removed["menus"]))
        print("")
        print(summarize_items("Removed from r_instat_not_menus.json (text: rules)", removed["notMenus"]))
        print("")

    if (
        len(classified_missing) == 0
        and not removed["menus"]
        and not removed["notMenus"]
        and not auto_removed["menus"]
        and not auto_removed["notMenus"]
    ):
        print("Nothing to sync - all extracted strings exist in EN JSON.")
        return 0

    added = append_missing_strings(base_dir, classified_missing)

    print("Wrote EN JSON updates.")
    print("")
    print(summarize_items("Appended to r_instat_menus.json", added["menus"]))
    print("")
    print(summarize_items("Appended to r_instat_not_menus.json", added["notMenus"]))
    print("")
    print("Next: review the diff and commit the EN JSON changes along with your VB edits.")

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
